src/main/java/com/coveo/pushapiclient/DocumentUploadQueue.java

```diff
@@ -1,21 +1,105 @@
+     * limit of 5MB.
+     *
+     * @param uploader The upload strategy to be used for document uploads.
+     */
     public DocumentUploadQueue(UpdloadStrategy uploader) {
+        this.documentToAddList = new ArrayList<>();
+        this.documentToDeleteList = new ArrayList<>();
         this.uploader = uploader;
     }
 
+    /**
+     * Flushes the accumulated documents by applying the upload strategy.
+     *
+     * @throws IOException          If an I/O error occurs during the upload.
+     * @throws InterruptedException If the upload process is interrupted.
+     */
     public void flush() throws IOException, InterruptedException {
-        throw new UnsupportedOperationException("Unimplemented method (TODO: LENS-856)");
+        if (this.isEmpty()) {
+            return;
+        }
+        BatchUpdate batch = this.getBatch();
+        // TODO: LENS-871: support concurrent requests
+        this.uploader.apply(batch);
+        this.size = 0;
+        this.documentToAddList.clear();
+        this.documentToDeleteList.clear();
     }
 
+    /**
+     * Adds a {@link DocumentBuilder} to the upload queue and flushes the queue if
+     * it exceeds the maximum content length.
+     * See {@link DocumentUploadQueue#flush}.
+     *
+     * @param document The document to be added to the index.
+     * @throws IOException          If an I/O error occurs during the upload.
+     * @throws InterruptedException If the upload process is interrupted.
+     */
     public void add(DocumentBuilder document) throws IOException, InterruptedException {
-        throw new UnsupportedOperationException("Unimplemented method (TODO: LENS-856)");
+        if (document == null) {
+            return;
+        }
+
+        final int sizeOfDoc = document.marshal().getBytes().length;
+        if (this.size + sizeOfDoc >= this.maxQueueSize) {
+            this.flush();
+        }
+        if (document != null) {
+            documentToAddList.add(document);
+            this.size += sizeOfDoc;
+        }
+    }
+
+    /**
+     * Adds a {@link DeleteDocument} to the upload queue and flushes the queue if
+     * it exceeds the maximum content length.
+     * See {@link DocumentUploadQueue#flush}.
+     *
+     * @param document The document to be delete from the index.
+     * @throws IOException          If an I/O error occurs during the upload.
+     * @throws InterruptedException If the upload process is interrupted.
+     */
+    public void add(DeleteDocument document) throws IOException, InterruptedException {
+        if (document == null) {
+            return;
+        }
+
+        final int sizeOfDoc = document.marshalJsonObject().toString().getBytes().length;
+        if (this.size + sizeOfDoc >= this.maxQueueSize) {
+            this.flush();
+        }
+        if (document != null) {
+            documentToDeleteList.add(document);
+            this.size += sizeOfDoc;
```

Comment:
same logic as for the add , maybe we can refactor inside a function and hide the increase of the size ?